Add launch daemon plist for VNet

ravicious opened the pull request:
https://github.com/gravitational/teleport/assets/27113/43564260-1c1c-49fa-9f6a-553d9dace2e9

```
VNET_DAEMON=1 $skel/tsh.app/Contents/MacOS/tsh vnet -d
```

This PR makes it so running the above command using [a signed version of tsh](https://github.com/gravitational/teleport/blob/master/build.assets/macos/tshdev/README.md) adds a login item for VNet's daemon.

The login item functions as a launch daemon that is started on demand through XPC. The name of the XPC service is advertised in the plist file under `MachServices`. Since this PR doesn't include any code which calls this XPC service, the daemon won't be spawned. The code is safe to compile and run, after trashing and removing tsh.app it shouldn't leave any residue in the system.

The actual implementation of the daemon will be added in a subsequent PR, along with code which calls the XPC service.

## How the launch daemon works

The launch daemon is added through [`SMAppService`](https://developer.apple.com/documentation/servicemanagement/smappservice?language=objc) available in macOS 13.0+. [The RFD wasn't updated yet](https://github.com/gravitational/teleport/blob/master/rfd/0163-vnet.md#daemon) to reflect how the daemon is going to actually run, but it's not as important for this PR specifically. I'll update it in the upcoming PR.

So far we've been using `osascript` to spawn `tsh vnet-admin-setup` as root. This command then configures the OS for VNet and checks `TELEPORT_HOME` ever 10 seconds to update the DNS configuration. The daemon is going to execute the same code as that command, with the difference being that the daemon will be spawned by an unprivileged tsh sending a message through XPC.

When the user first launches VNet, a login item is added. Before calling the XPC service, we have to wait for the login item to be enabled by the user (since it requires root privileges). In theory, a system admin could add the plist of VNet daemon to launchd. This could work with MDMs, but I haven't actually checked if it works.

Since `SMAppService` works only with app bundles, we default to the old osascript method. This way Teleport developers can run VNet without having to go through the app signing flow, which is a bit complicated to set up.

ravicious committed:
Detect VNet daemon feature flag and signing status

ravicious committed:
Register daemon plist

ravicious committed:
Wait for enablement of login item

ravicious commented on lib/vnet/daemon/client_darwin.go:
```diff
@@ -0,0 +1,74 @@
+// You should have received a copy of the GNU Affero General Public License
+// along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+//go:build darwin
+// +build darwin
+
+package daemon
+
+// #cgo CFLAGS: -Wall -xobjective-c -fblocks -fobjc-arc -mmacosx-version-min=10.15
+// #cgo LDFLAGS: -framework Foundation -framework ServiceManagement
+// #include "client_darwin.h"
+import "C"
```
Although within this PR the `daemon` package is only ever called from code compiled on darwin only, in the future PRs I'm going to add some common Go code that defines structs used on other platforms as well (it's used only in type signatures there).

Hence why every file, both Go and Obj-C, has `_darwin` in it. cgo on Linux doesn't let you build a package if it includes Obj-C files.

codingllama replied:
I added build tags to the .m files on touchid (and similar packages). That works too.

- https://github.com/gravitational/teleport/blob/7310b2f17ac0fe643fb1552ba2e06b395090a10e/lib/auth/touchid/authenticate.m#L1-L2

ravicious commented on lib/vnet/daemon/client_darwin.m:
```diff
@@ -0,0 +1,99 @@
+}
+
+void RegisterDaemon(struct RegisterDaemonResult *result) {
+    if (@available(macOS 13, *)) {
```
I'm yet to actually verify that those availability checks work on macOS < 13.0, but I have that in my TODO list.

Without them, there's a warning emitted by the compiler which tells you that you ought to wrap calls to `SMAppService` in `@available`. Unfortunately, it seems to be pretty basic and doing an early return doesn't work – the code that's available only on certain versions must be wrapped within an `if (@available(…))`.

I don't think we should be bumping our minimal supported version of macOS just due to a single feature that needs 13.0.

ravicious commented on lib/vnet/daemon/client_darwin.go:
```diff
@@ -0,0 +1,74 @@
+		return false, nil
+	}
+
+	if err := exec.CommandContext(ctx, "codesign", "--verify", bundlePath).Run(); err != nil {
```
`os.Executable` wouldn't work here – it has to be the path to the bundle itself, otherwise `codesign --verify` returns what I'd consider false positives. For example, if you run it on `build/tsh`, it tells you that everything is A-okay. And that's true – `build/tsh` is not in any app bundle, which means there's no signing requirements placed upon it, so technically it fulfills all requirements.

ravicious replied:
`codesign --verify` doesn't work when tsh is executed from a symlink in `/usr/local/bin`. This can happen when tsh is installed through tsh.app.

In that scenario, `C.BundlePath` returns `/usr/local/bin`, which has a similar effect as calling `codesign --verify` on `build` in the teleport repo.

I'll need to switch to `spctl --assess`, which AFAIR works with individual items. So instead of finding the bundle path, we'll be able to call it directly on the tsh executable from `Contents/MacOS`.

ravicious replied:
`spctl --assess` fails on unnotarized bundles, which means running it on `tshdev/tsh.app/Contents/MacOS/tsh` that's signed locally won't work.

ravicious replied:
I think I'll keep using `codesign --verify`, but instead of using Obj-C to get the bundle path, I'll resolve any symlinks to `os.Executable`, then check if the dir path ends with `Contents/MacOS` and then just remove that suffix.

This check isn't done for security, it's merely to know whether we're running from inside a signed package or if we should run the osascript prompt.

ravicious commented:
@codingllama I'm adding you for your Obj-C expertise and @nklaassen for your VNet knowledge.

ravicious committed:
Add dev docs on how to work with launch daemons

ravicious commented on lib/vnet/daemon/client_darwin.m:
```diff
@@ -0,0 +1,99 @@
+        return bundleIdentifier;
+    }
+    
+    return [NSString stringWithFormat:@"%@.vnetd", bundleIdentifier];
```
I might need to adjust assumptions made about bundle identifiers here. This code works locally, but I just made a tag build and it doesn't work.

I assumed that in development, our bundle identifier is just `com.goteleport.tshdev` (as listed in Info.plist). However, it's actually `K497G57PDJ.com.goteleport.tshdev`.

I haven't checked yet if this is the root cause of the problem, since Connect simply hangs either when calling `C.DaemonStatus` or `C.RegisterDaemon`.

ravicious replied:
Ok, so my assumptions were correct and it was an issue with cache. For some reason, macOS really doesn't like when I switch from a local signed build to a tag build. Weird, because both are signed with the same key.

`sfltool dumpbtm` returns this, hence why I think my assumptions about identifiers are correct. The bundle identifier _is_ `com.goteleport.tshdev`, so I think it's reasonable for the daemon plist to be identified as `com.goteleport.tshdev.vnetd`.

```
 #5:
                 UUID: D07CBD74-9A97-479C-986F-0C7644E0F5C9
                 Name: tsh
       Developer Name: Ada Lin
      Team Identifier: K497G57PDJ
                 Type: app (0x2)
          Disposition: [disabled, allowed, visible, notified] (10)
           Identifier: identifier "K497G57PDJ.com.goteleport.tshdev" and anchor apple generic and certificate 1[field.1.2.840.113635.100.6.2.6] /* exists */ and certificate leaf[field.1.2.840.113635.100.6.1.13] /* exists */ and certificate leaf[subject.OU] = K497G57PDJ
                  URL: file:///Applications/tsh.app/
           Generation: 0
    Bundle Identifier: com.goteleport.tshdev
  Embedded Item Identifiers:
    #1: com.goteleport.tshdev.vnetd

 #6:
                 UUID: 77943637-3F1F-429F-B118-E70D207D2AAA
                 Name: tsh
       Developer Name: (null)
                 Type: daemon (0x10)
          Disposition: [enabled, allowed, visible, notified] (11)
           Identifier: com.goteleport.tshdev.vnetd
                  URL: Contents/Library/LaunchDaemons/com.goteleport.tshdev.vnetd.plist
      Executable Path: Contents/MacOS/tsh
           Generation: 1
    Parent Identifier: identifier "K497G57PDJ.com.goteleport.tshdev" and anchor apple generic and certificate 1[field.1.2.840.113635.100.6.2.6] /* exists */ and certificate leaf[field.1.2.840.113635.100.6.1.13] /* exists */ and certificate leaf[subject.OU] = K497G57PDJ
```

codingllama commented on lib/vnet/daemon/client_darwin.go:
```diff
@@ -0,0 +1,197 @@
+}
+
+func register(ctx context.Context) (serviceStatus, error) {
+	var result C.RegisterDaemonResult
+
+	C.RegisterDaemon(&result)
+	defer func() {
+		C.free(unsafe.Pointer(result.error_description))
+	}()
```
nit: Keep the defer next to the declaration.

```suggestion
	var result C.RegisterDaemonResult
	defer func() {
		C.free(unsafe.Pointer(result.error_description))
	}()

	C.RegisterDaemon(&result)
```

codingllama commented on lib/vnet/daemon/client_darwin.go:
```diff
@@ -0,0 +1,197 @@
+
+	if !result.ok {
+		status := serviceStatus(result.service_status)
+		// Docs for registerAndReturnError [1] don't seem to cover a situation in which the user adds
```
nit: You have to "anchor" the link to something:

```suggestion
		// Docs for [registerAndReturnError][1] don't seem to cover a situation in which the user adds
```

codingllama commented on lib/vnet/daemon/client_darwin.go:
```diff
@@ -0,0 +1,197 @@
+		status := serviceStatus(result.service_status)
+		// Docs for registerAndReturnError [1] don't seem to cover a situation in which the user adds
+		// the launch daemon for the first time. In that case, that method returns "Operation not permitted"
+		// error (Code=1, Domain=SMAppServiceErrorDomain). That error doesn't seem to exist in Service
+		// Management Errors [2]. However, all this means is that the launch daemon was added to the
```
```suggestion
		// error (Code=1, Domain=SMAppServiceErrorDomain). That error doesn't seem to exist in [Service
		// Management Errors][2]. However, all this means is that the launch daemon was added to the
```

codingllama commented on lib/vnet/daemon/client_darwin.go:
```diff
@@ -0,0 +1,197 @@
+		//
+		// We can confirm this by looking at the returned service status.
+		//
+		// [1] https://developer.apple.com/documentation/servicemanagement/smappservice/register()?language=objc
+		// [2] https://developer.apple.com/documentation/servicemanagement/service-management-errors?language=objc
```
nit: Improper link syntax:

```suggestion
		// [1]: https://developer.apple.com/documentation/servicemanagement/smappservice/register()?language=objc
		// [2]: https://developer.apple.com/documentation/servicemanagement/service-management-errors?language=objc
```

codingllama commented on lib/vnet/daemon/client_darwin.go:
```diff
@@ -0,0 +1,197 @@
+
+// https://developer.apple.com/documentation/servicemanagement/smappservice/status-swift.enum?language=objc
+const (
+	serviceStatusNotRegistered    serviceStatus = 0
+	serviceStatusEnabled          serviceStatus = 1
+	serviceStatusRequiresApproval serviceStatus = 2
+	serviceStatusNotFound         serviceStatus = 3
```
Suggestion:

```suggestion
	serviceStatusNotRegistered serviceStatus = iota
	serviceStatusEnabled
	serviceStatusRequiresApproval
	serviceStatusNotFound
```

codingllama commented on lib/vnet/daemon/client_darwin.go:
```diff
@@ -0,0 +1,197 @@
+	case serviceStatusNotFound:
+		return "not found"
+	default:
+		return fmt.Sprintf("%d", int(s))
```
```suggestion
		return strconv.Itoa(int(s))
```

Prefer strconv to Sprintf conversions if you can.

codingllama commented on lib/vnet/daemon/client_darwin.h:
```diff
@@ -0,0 +1,53 @@
+// bundlePath is an empty string if the bundle details could not be fetched.
+// It might return a path even for executables that are not in a bundle.
+// In that case, calling codesign --verify on that path will simply return with 1.
+void BundlePath(struct BundlePathResult *result);
```
Suggestion: use the typedef declared above, explicitly mark output params:

```suggestion
void BundlePath(BundlePathResult *outResult);
```

ravicious replied:
Could you tell me or give me a quick link to something that explains the difference between `void BundlePath(struct BundlePathResult *result)` and `void BundlePath(BundlePathResult *result)`? I wasn't aware of the latter.

codingllama replied:
Sure.

My poor man's explanation is this: structs in C have their own "namespace", so when you declare "struct Foo", its "type" is always "struct Foo" (instead of just "Foo", which is what we've come to expect from other languages).

typedefs are a way to give a nicer, concise name to complex things. Because typing "struct Foo" everywhere is cumbersome, and forces your program to know that Foo is a struct, it's common to always typedef structs to a "nicer" name so we can drop the "struct" part from the syntax.

Usually this means we typedef "struct Foo" to "Foo" and get own with our lives. In some codebases you'll see "struct _Foo" typedef as "Foo" as well.

- https://stackoverflow.com/a/2566035

codingllama commented on lib/vnet/daemon/client_darwin.h:
```diff
@@ -0,0 +1,53 @@
+    const char * bundlePath;
+} BundlePathResult;
+
+// BundlePath updates bundlePath of result with the path
```
Document that the caller is responsible for freeing the BundlePathResult strings.

codingllama commented on lib/vnet/daemon/client_darwin.h:
```diff
@@ -0,0 +1,53 @@
+
+typedef struct RegisterDaemonResult {
+    bool ok;
+    // service_status is fetched even if [service registerAndReturnError] fails,
```
nit: I don't understand what the "markdown" on `[service registerAndReturnError]` is supposed to mean.

ravicious replied:
It's supposed to be Obj-C syntax for calling a method. 😎

codingllama replied:
Oh, of course.

codingllama commented on lib/vnet/daemon/client_darwin.h:
```diff
@@ -0,0 +1,53 @@
+    // service_status is fetched even if [service registerAndReturnError] fails,
+    // for debugging purposes.
+    int service_status;
+    const char * error_description;
```
nit:

```suggestion
    const char *error_description;
```

codingllama commented on lib/vnet/daemon/client_darwin.h:
```diff
@@ -0,0 +1,53 @@
+#import <Foundation/Foundation.h>
+
+typedef struct BundlePathResult {
+    const char * bundlePath;
```
nit:

```suggestion
    const char *bundlePath;
```

codingllama commented on lib/vnet/daemon/client_darwin.h:
```diff
@@ -0,0 +1,53 @@
+    bool ok;
+    // service_status is fetched even if [service registerAndReturnError] fails,
+    // for debugging purposes.
+    int service_status;
```
nit: Lower camel (bundlePath) or lower_snake (service_status)?  :P

ravicious replied:
I'm going to stick to lower_snake for fields. `bundlePath` was one that I forgot to update. ;)

codingllama commented on lib/vnet/daemon/client_darwin.h:
```diff
@@ -0,0 +1,53 @@
+// RegisterDaemon attempts to register the daemon. After the registration attempt,
+// it fetches the daemon status.
+// Pretty much a noop if the daemon is already registered and enabled.
+void RegisterDaemon(struct RegisterDaemonResult *result);
```
Take advantage of the typedefs:

```suggestion
void RegisterDaemon(RegisterDaemonResult *result);
```

codingllama commented on lib/vnet/daemon/client_darwin.h:
```diff
@@ -0,0 +1,53 @@
+// RegisterDaemon attempts to register the daemon. After the registration attempt,
+// it fetches the daemon status.
+// Pretty much a noop if the daemon is already registered and enabled.
+void RegisterDaemon(struct RegisterDaemonResult *result);
```
Should this function return some sort of success indicator?

`int RegisterDaemon(RegisterDaemonResult *result)`, perhaps?

ravicious replied:
`RegisterDaemonResult` already has the `bool ok` field. I swear I thought I lifted this pattern some `lib/auth/touchid` code, but now that I looked at it, it doesn't actually use this pattern at all.

Should I then make it so that it returns `0` on success and `1` on error?

codingllama replied:
The ok field works too, you are right. Maybe mention that the caller should check it on the docs?

If you'd rather to change to 0=success, non-zero=error, that's also fine by me.

codingllama commented on lib/vnet/daemon/client_darwin.h:
```diff
@@ -0,0 +1,53 @@
+void RegisterDaemon(struct RegisterDaemonResult *result);
+
+// DaemonStatus returns the current status of the daemon's service in SMAppService.
+// Returns -1 if the given macOS version doesn't support SMAppService.
```
Is zero a success? Do any of the values mean the daemon is running? Please document.

codingllama commented on lib/vnet/daemon/client_darwin.h:
```diff
@@ -0,0 +1,53 @@
+// Returns -1 if the given macOS version doesn't support SMAppService.
+int DaemonStatus(void);
+
+void OpenSystemSettingsLoginItems(void);
```
Please document.

Should this method return some sort of success indicator?

codingllama commented on lib/vnet/daemon/client_darwin.h:
```diff
@@ -0,0 +1,53 @@
+
+// VNECopyNSString duplicates an NSString into an UTF-8 encoded C string.
+// The caller is expected to free the returned pointer.
+char *VNECopyNSString(NSString *val);
```
Ditch this from the header and move into client_darwin.m? It's best if Go code doesn't touch NS types, imo.

codingllama commented on lib/vnet/daemon/client_darwin.h:
```diff
@@ -0,0 +1,53 @@
+#ifndef client_darwin_h
+#define client_darwin_h
```
FYI, I tend to use Google style for these:

- https://google.github.io/styleguide/cppguide.html#The__define_Guard

TBH I was intending to keep to Google style for most of the native code, but I think I might have strayed at some point.  🤷

ravicious replied:
Thanks, I wasn't aware that there are different styles for these guards. These were automatically generated by Xcode.

codingllama replied:
There are conflicting styles for everything here. Whatever you prefer works.

There's also `#pragma once`, but I don't remember why I didn't do that in the first place in these parts. Probably because following Google Style seemed simpler.

codingllama commented on lib/vnet/daemon/client_darwin.go:
```diff
@@ -0,0 +1,197 @@
+// You should have received a copy of the GNU Affero General Public License
+// along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+//go:build darwin
+// +build darwin
+
```
FYI: files that end in _darwin are automatically tagged as "// go:build darwin". Same for other platforms.

```suggestion
```

ravicious replied:
Oh thanks. A long time ago I was actually wondering why it needs to be that way, I couldn't find a reason so I assumed it was for some kind of backwards compatibility, but it's not the case I see.

codingllama replied:
The duplicate `//go:` and ``// +` directives were for backwards compat, although I think that has fallen by now. The _platform suffixes have been around for a long time afaik.

codingllama commented on lib/vnet/daemon/client_darwin.m:
```diff
@@ -0,0 +1,99 @@
+// Teleport
```
Repeat the Go build tags here so you don't get other platforms trying to compile this code.

```suggestion
//go:build darwin
// +build darwin

// Teleport
```

ravicious replied:
Why are those repeated build tags needed in `_darwin.m` files but not in `_darwin.go` files? From what I tested, `_darwin.m` was enough to avoid compiling this on other platforms. Without the `_darwin` part, cgo on Linux wouldn't let me compile `lib/vnet/daemon`, even if there was no Go file on Linux that included this Obj-C file.

codingllama replied:
I wasn't sure about .m, but if it works feel free to resolve the comment.

codingllama commented on lib/vnet/daemon/client_darwin.m:
```diff
@@ -0,0 +1,99 @@
+//
+// You should have received a copy of the GNU Affero General Public License
+// along with this program.  If not, see <http://www.gnu.org/licenses/>.
+#import <Foundation/Foundation.h>
+#import <ServiceManagement/ServiceManagement.h>
+
+#include "client_darwin.h"
```
I tend to follow Google Style here too, so I ended up something like:

- own header import
- frameworks
- stdlib
- other system imports
- custom headers

```suggestion

#include "client_darwin.h"

#import <Foundation/Foundation.h>
#import <ServiceManagement/ServiceManagement.h>
```

- https://google.github.io/styleguide/cppguide.html#Names_and_Order_of_Includes

codingllama commented on lib/vnet/daemon/client_darwin.m:
```diff
@@ -0,0 +1,99 @@
+        return bundleIdentifier;
+    }
+    
+    return [NSString stringWithFormat:@"%@.vnetd", bundleIdentifier];
```
Could we get away with the app not knowing its own bundle? I think that would be best, if we can.

ravicious replied:
I'll think about it, but off the top of my head I think that won't be possible. The daemon needs to have a label that's advertised in the plist _and_ we want to use different labels between tshdev and tsh, just as we do with the bundle identifier of tsh.app.

I wanted to avoid injecting the same identifier twice, e.g. once in plist and another time through build tags just so that `client_darwin.go` get the identifier without consulting `[[NSBundle mainBundle] bundleIdentifier]`.

codingllama replied:
If we need it in the bundle then querying NSBundle seems for the best. +1.

codingllama commented on lib/vnet/daemon/client_darwin.m:
```diff
@@ -0,0 +1,99 @@
+        // [service registerAndReturnError] succeeded or failed.
+        result->service_status = (int) service.status;
+    } else {
+        result->error_description = strdup("Service Management APIs are available on macOS 13.0+");
```
```suggestion
        result->error_description = strdup("Service Management APIs only available on macOS 13.0+");
```

codingllama commented on build.assets/macos/tshdev/README.md:
```diff
@@ -36,6 +36,8 @@ $skel/sign.sh tsh                  # sign tsh binary
 security find-identity -vp codesigning      # list codesign identities
 security cms -D -i /path/to/profile         # inspect provisioning profile
 codesign -d --entitlements /path/to/binary  # shows binary entitlements
+codesign --verify --verbose                 # verifies code signatures
+codesign -dv --verbose=4                    # verifies code signatures with verbose output
```
I tend to use `codesign -dvv`, I think it might be the same?

Include only the more verbose one here?

codingllama commented on build.assets/macos/tshdev/README.md:
```diff
@@ -79,3 +81,104 @@ through the steps here unless you are trying to recreate parts of the skeleton
     For example, click "build" and copy the app using "Product -> Show Build
     Folder in Finder".
+
+## Working with launch daemons
```
Will every execution of a signed tsh register the login item?

What do I do if I don't want that to happen?

ravicious replied:
No, the login item is registered only when `tsh vnet` is ran, because that's the only code path that [registers the daemon](https://developer.apple.com/documentation/servicemanagement/smappservice/register()?language=objc). Any other execution of a signed tsh is not going to register the login item.

Something to remember in the future is that enabling one login item from a bundle enables all of them. I saw that mentioned in one of those Apple dev forum threads, but I forgot to capture a link.

I suppose this also addresses the second question, doesn't it?

I guess one improvement I could make is to skip the call to register if the service is already enabled.

codingllama replied:
That clears some of it up, thanks.

I think it's worth having a way to skip the registration for dev builds, even if it's just for `tsh vnet`.

> Something to remember in the future is that enabling one login item from a bundle enables all of them. I saw that mentioned in one of those Apple dev forum threads, but I forgot to capture a link.

That's gonna be a surprise if it happens, for sure. Thanks for flagging it now.

> I guess one improvement I could make is to skip the call to register if the service is already enabled.

If attempting to register is cheap/fast then I don't think it's much of a problem, assuming I understand this correctly.

codingllama replied:
> I think it's worth having a way to skip the registration for dev builds, even if it's just for tsh vnet.

Nevermind this part here, https://github.com/gravitational/teleport/pull/43746#discussion_r1666079146 has a better thread on this.

codingllama commented on build.assets/macos/tshdev/README.md:
```diff
@@ -79,3 +81,104 @@ through the steps here unless you are trying to recreate parts of the skeleton
+basically wipes the whole database of Launch Services, meaning that the status of all login items
+gets reset – this includes other software installed on your machine.
+
+After running it, you must restart the device. We found it to be the most helpful when switching
+between local signed builds and tag builds. For some reason, in that scenario macOS wouldn't want to
```
Could this be because of the change of .app location, or the fact that it gets confused because now multiple .apps for the same bundle exists?

I've run into weirdness with that before, too.

ravicious replied:
In that case AFAIR I was putting tsh.app in the same location and there was no other bundles – I was doing that in a macOS VM. But maybe I did miss something. I couldn't find a good way to establish a link between a bundle identifier and the actual app bundle on disk that launchd is going to use for that particular login item, unfortunately.

codingllama commented on lib/vnet/daemon/client_darwin.h:
```diff
@@ -0,0 +1,53 @@
+// In that case, calling codesign --verify on that path will simply return with 1.
+void BundlePath(struct BundlePathResult *result);
+
+// Returns the label for the daemon by getting the identifier of the bundle
+// this executable is shipped in and appending ".vnetd" to it.
+//
+// The returned string might be empty if the executable is not in a bundle.
+//
+// The filename and the value of the Label key in the plist file and the Mach
+// service of of the daemon must match the string returned from this function.
+NSString * DaemonLabel(void);
+
+// DaemonPlist takes the result of DaemonLabel and appends ".plist" to it
+// if not empty.
+NSString * DaemonPlist(void);
```
Move DaemonLabel and DaemonPlist from the .h to the .m. We don't want Go code touching any NS-types, memory management becomes anyone's guess at that point.

We can keep in the header only types and functions Go cares about (since you only have one .m file).

codingllama commented on lib/vnet/daemon/client_darwin.go:
My high-level comment here is that I would be able to swap, on build time, whether I want no registration, osascript-based registration or SMAppService-based registration (likely defaulting to no registration).

This lets the binary be less self-aware, which kills some code and ifs we have. It also makes prod binaries follow a single path.

WDYT?

codingllama replied:
I would also consider cutting the osascript method, although it does add value today. (Less so when we get the installer situation sorted.)

codingllama replied:
I thought the osascript approach was another way to make a launch daemon.  🤦 

It looks like the choice is only between osascript vs SMAppService, then. I still think it's worth making the choice at compile time, likely with osascript as the default so we make dev simpler.

codingllama reviewed (commented):
Seems reasonable to me. I don't have much experience with launch daemons (apart from some local scripts I have), so sorry I can't help more.

ravicious commented on lib/vnet/daemon/client_darwin.go:
```diff
@@ -0,0 +1,197 @@
+	// going to show the notification about a new login item. Instead, we just open the login items
+	// ourselves to direct the user towards enabling the login item.
+	if initialStatus == serviceStatusRequiresApproval {
+		C.OpenSystemSettingsLoginItems()
```
Just opening the menu is enough.

I was planning to show an in-app notification in Connect. In tsh, we can certainly print the same text to stdout, asking the user to enable the login item for tsh.app.

The in-app notification will be added in a separate PR, but I will make tsh print a message in this PR.

ravicious committed:
Run code through autoformatter

https://formatter.org/objc-formatter
https://google.github.io/styleguide/objcguide.html#spacing-and-formatting

Coding Style: Google
Indent Width: 2
Column Limit: 100

ravicious committed:
Remove unnecessary darwin build tags

ravicious committed:
Decide to use launch daemon at build time

ravicious committed:
Address minor feedback

ravicious committed:
Show a message when waiting for bg item to be enabled

ravicious commented on lib/vnet/daemon/client_darwin.m:
```diff
@@ -0,0 +1,98 @@
+// The filename and the value of the Label key in the plist file and the Mach
+// service of of the daemon must match the string returned from this function.
+NSString *DaemonLabel(void) {
+  NSBundle *main = [NSBundle mainBundle];
```
I moved the decision behind whether to use a launch daemon or osascript behind a build tag. This removed the need to perform those weird shenanigans with `codesign --verify` in Go.

However, that's not the end of the issues. It turns out that `[NSBundle mainBundle]` doesn't handle symlinks well either. When executing `/usr/local/bin/tsh vnet` with the tsh symlinked to `/Applications/tsh.app/Contents/MacOS/tsh`, it thinks that the bundle is at `/usr/local/bin`, which is not the case.

I'm going to add one more change. In Go, I'm going to call `os.Executable`, follow any symlinks, and then check if the dir path ends with `Contents/MacOS`. If so, I'm going to pass that as `bundle_path` to `C.RegisterDaemon`. Then I'm going to call [`[NSBundle bundleWithPath]`](https://developer.apple.com/documentation/foundation/nsbundle/1495012-bundlewithpath).

ravicious commented on Makefile:
```diff
@@ -178,6 +178,15 @@ TOUCHID_MESSAGE := with-Touch-ID
+VNETDAEMON_MESSAGE := without-VNet-daemon
+ifeq ("$(VNETDAEMON)", "yes")
+VNETDAEMON_MESSAGE := with-VNet-daemon
+VNETDAEMON_TAG := vnetdaemon
```
I have [a branch with updates to teleport.e](https://github.com/gravitational/teleport.e/tree/r7s/vnet-1-plist) ready to go as well. I wanted to wait with creating a PR until I get an approval here, as GitHub is going to assign code owners for a review, even if I create the PR as a draft.

codingllama commented on Makefile:
```diff
@@ -828,7 +837,7 @@ test-go-prepare: ensure-webassets bpf-bytecode $(TEST_LOG_DIR) ensure-gotestsum
 test-go-unit: SUBJECT ?= $(shell go list ./... | grep -vE 'teleport/(e2e|integration|tool/tsh|integrations/operator|integrations/access|integrations/lib)')
 test-go-unit:
-	$(CGOFLAG) go test -cover -json -tags "$(PAM_TAG) $(FIPS_TAG) $(BPF_TAG) $(LIBFIDO2_TEST_TAG) $(TOUCHID_TAG) $(PIV_TEST_TAG)" $(PACKAGES) $(SUBJECT) $(FLAGS) $(ADDFLAGS) \
+	$(CGOFLAG) go test -cover -json -tags "$(PAM_TAG) $(FIPS_TAG) $(BPF_TAG) $(LIBFIDO2_TEST_TAG) $(TOUCHID_TAG) $(PIV_TEST_TAG) $(VNETDAEMON_TAG)" $(PACKAGES) $(SUBJECT) $(FLAGS) $(ADDFLAGS) \
```
You might wanna test both with and without the tag, so one of the code paths doesn't break. (Do the builders even test macOS specific code?)

ravicious replied:
I think the builders don't run macOS code, but I'm going to add an equivalent of `test-go-touch-id` in case someone runs tests locally on a Mac.

codingllama commented on lib/vnet/daemon/client_darwin.m:
```diff
@@ -0,0 +1,98 @@
+
+#import <Foundation/Foundation.h>
+#import <ServiceManagement/ServiceManagement.h>
+
```
For `strdup`. (Depending on indirect includes might break you if they ever change.)

```suggestion

#include <string.h>

```

- https://google.github.io/styleguide/cppguide.html#Include_What_You_Use

codingllama commented on lib/vnet/daemon/client_darwin.go:
```diff
@@ -0,0 +1,163 @@
+
+// #cgo CFLAGS: -Wall -xobjective-c -fblocks -fobjc-arc -mmacosx-version-min=10.15
+// #cgo LDFLAGS: -framework Foundation -framework ServiceManagement
+// #include "client_darwin.h"
```
For the `free` function:

```suggestion
// #include <stdlib.h>
// #include "client_darwin.h"
```

- https://google.github.io/styleguide/cppguide.html#Include_What_You_Use

codingllama commented on lib/vnet/daemon/client_darwin.m:
```diff
@@ -0,0 +1,98 @@
+
+// VNECopyNSString duplicates an NSString into an UTF-8 encoded C string.
+// The caller is expected to free the returned pointer.
+char *VNECopyNSString(NSString *val) {
```
nit:

```suggestion
const char *VNECopyNSString(NSString *val) {
```

codingllama reviewed (commented):
LGTM.

Let us know when you are done making changes.

ravicious committed:
Correctly resolve bundle path if executable is symlinked

mainBundle of NSBundle would not resolve symlinks, resulting in a wrong
bundle information when tsh was symlinked to /usr/local/bin/tsh by tsh.pkg.

nklaassen commented on build.assets/macos/tsh/tsh.app/Contents/Library/LaunchDaemons/com.gravitational.teleport.tsh.vnetd.plist:
```diff
@@ -0,0 +1,24 @@
+<!DOCTYPE plist PUBLIC "-//Apple//DTD PLIST 1.0//EN" "http://www.apple.com/DTDs/PropertyList-1.0.dtd">
+<plist version="1.0">
+<dict>
+	<key>Label</key>
+	<string>com.gravitational.teleport.tsh.vnetd</string>
+	<key>BundleProgram</key>
+	<string>Contents/MacOS/tsh</string>
+  <key>ProgramArguments</key>
+  <array>
+      <string>tsh</string>
+      <string>vnet-daemon</string>
+  </array>
+	<key>MachServices</key>
+	<dict>
+    <key>com.gravitational.teleport.tsh.vnetd</key>
+		<true/>
+	</dict>
+  <key>StandardErrorPath</key>
+  <string>/var/log/vnet.log</string>
+  <key>StandardOutPath</key>
+  <string>/var/log/vnet.log</string>
```
looks like inconsistent tabs vs spaces here

ravicious committed:
Run tests for untagged vnetdaemon code

nklaassen commented on build.assets/macos/tshdev/README.md:
````diff
@@ -79,3 +80,116 @@ through the steps here unless you are trying to recreate parts of the skeleton
+
+When there's no login item for a daemon and an app attempts to register it for the first time, macOS
+shows a notification about it which says "tsh.app added items that can run in the background for all
+users. Do you want to allow this?". When working on a launch daemon, at some point macOS is going to
+stop showing the notification. In Console.app, you're going to see this:
+
+```
+Exceeded max notifications for tsh. Please file a Feedback report or contact the developer.
+item=uuid=4DF802A6-A9DE-4EF8-A6FB-4B046E50E1DE, name=tsh, type=daemon, disposition=[enabled,
+disallowed, visible, not notified], identifier=com.goteleport.tshdev.vnetd,
+url=Contents/Library/LaunchDaemons/com.goteleport.tshdev.vnetd.plist -- file:///
+```
````
what do you have to do at this point? could users hit this if they ignore the notification enough times? can we warn the user that this notification is going to come up and they need to click allow?

ravicious replied:
If a notification doesn't get displayed, you pretty much just have to go to the login items yourself.

I think users are unlikely to hit this. Someone would have to move Connect/tsh to the trash, then empty the trash, go to Login Items, then install Connect/tsh again and start VNet and do it multiple times during the day. If you just trash Connect/tsh and then install it again without opening Login Items, then the background item stays enabled, I believe.

In any case, it's quite easy to miss the notification even when it's displayed, especially when using multiple monitors. I did it a couple of times when working on this PR. To help with this, I already added a message that's displayed by `tsh vnet` when we see that the daemon is waiting for an approval:

> To start VNet, please enable the background item for tsh.app in the Login Items section of System Settings.
> Waiting for the background item to be enabled…

Later I'm going to add an in-app notification to Connect which shows a similar message. This is especially important for Connect since the system notification looks like this:

<img width="367" alt="background-item" src="https://github.com/gravitational/teleport/assets/27113/cb718d3d-b2f0-40eb-97b5-9d48a364c5bd">

I think it's important for Connect to explicitly mention tsh.app, otherwise users could get suspicious and not understand why they got a notification for something called tsh.app while using Connect.

nklaassen commented on build.assets/macos/tshdev/tsh.app/Contents/Library/LaunchDaemons/com.goteleport.tshdev.vnetd.plist:
```diff
@@ -0,0 +1,24 @@
+<!DOCTYPE plist PUBLIC "-//Apple//DTD PLIST 1.0//EN" "http://www.apple.com/DTDs/PropertyList-1.0.dtd">
+<plist version="1.0">
+<dict>
+	<key>Label</key>
+	<string>com.goteleport.tshdev.vnetd</string>
+	<key>BundleProgram</key>
+	<string>Contents/MacOS/tsh</string>
+  <key>ProgramArguments</key>
+  <array>
+      <string>tsh</string>
+      <string>vnet-daemon</string>
+  </array>
+	<key>MachServices</key>
+	<dict>
+    <key>com.goteleport.tshdev.vnetd</key>
+		<true/>
+	</dict>
+  <key>StandardErrorPath</key>
+  <string>/var/log/vnet.log</string>
+  <key>StandardOutPath</key>
+  <string>/var/log/vnet.log</string>
```
inconsistent tabs/spaces?

ravicious committed:
Add includes for standard library headers

ravicious commented:
Ok, I pushed the final change with how the bundle path is resolved in 9ac8a74.

nklaassen commented on lib/vnet/daemon/client_darwin.go:
```diff
@@ -0,0 +1,163 @@
+
+type serviceStatus int
+
+// https://developer.apple.com/documentation/servicemanagement/smappservice/status-swift.enum?language=objc
+const (
+	serviceStatusNotRegistered serviceStatus = iota
+	serviceStatusEnabled
+	serviceStatusRequiresApproval
+	serviceStatusNotFound
+)
```
I'd rather give these explicit values than use iota if we are matching an external definition for them

codingllama replied:
This was me  😅 

Rafal, feel free to revert to your original version.

codingllama commented on lib/utils/fs.go:
```diff
@@ -94,11 +94,19 @@ func NormalizePath(path string, evaluateSymlinks bool) (string, error) {
+
+		// If the symlink is relative, filepath.EvalSymlinks returns a path relative to the current dir.
+		// Resolve it to an absolute path to stay true to the godoc.
+		s, err = filepath.Abs(s)
```
Could we add test coverage for this?

ravicious replied:
I misunderstood how [`filepath.EvalSymlinks`](https://pkg.go.dev/path/filepath#EvalSymlinks) works.

> EvalSymlinks returns the path name after the evaluation of any symbolic links. If path is relative the result will be relative to the current directory, unless one of the components is an absolute symbolic link. EvalSymlinks calls [Clean](https://pkg.go.dev/path/filepath#Clean) on the result.

Specifically, "If path is relative" is about the path passed to `EvalSymlinks`, not the path of the symlink. I'll remove the changes to `lib/utils.NormalizePath`.

codingllama commented on lib/vnet/daemon/client_darwin.m:
```diff
@@ -35,38 +35,37 @@
     NSError *error;
 
-    service = [SMAppService daemonServiceWithPlistName:(DaemonPlist())];
+    service = [SMAppService daemonServiceWithPlistName:(DaemonPlist(@(bundle_path)))];
```
Do the @(bundle_path) conversion assume UTF-8?

ravicious replied:
`@(bundle_path)` does assume UTF-8. Do I need to do anything extra because of that?

https://clang.llvm.org/docs/ObjectiveCLiterals.html#boxed-expressions
https://developer.apple.com/documentation/foundation/nsstring/1497379-stringwithutf8string

codingllama replied:
Nope, just checking. Thanks for the links.

codingllama reviewed (approved):
LGTM, nicely done.

ravicious committed:
Format plists in Xcode

ravicious committed:
Use explicit values for serviceStatus

ravicious committed:
Revert changes to lib/utils.NormalizePath
